src/app/pcasts/dao/listening_histories_dao.py

```diff
@@ -1,25 +1,38 @@
 from app.pcasts.dao import episodes_dao
 from . import *
 
-def create_or_update_listening_history(episode_id, user):
-  optional_episode = episodes_dao.get_episode(episode_id, user.id)
-  if optional_episode:
-    optional_listening_history = ListeningHistory.\
-      query.\
-      filter(
-          ListeningHistory.episode_id == episode_id,
-          ListeningHistory.user_id == user.id
-      ).first()
-
-    if optional_listening_history:
-      optional_listening_history.update_listening_time()
-      db_utils.db_session_commit()
-      return optional_listening_history
-
-    listening_history = ListeningHistory(episode_id=episode_id, user_id=user.id)
-    return db_utils.commit_model(listening_history)
-  else:
-    raise Exception('Invalid episode_id provided')
+def create_or_update_listening_histories(episode_update_info_map, user):
+  episode_ids = episode_update_info_map.keys()
+  episodes = episodes_dao.get_episodes(episode_ids, user.id)
+  listening_histories = \
+    ListeningHistory.query.filter(ListeningHistory.episode_id.in_(episode_ids),
+                                  ListeningHistory.user_id == user.id).all()
+  listening_histories_by_episode = \
+    {lh.episode_id: lh for lh in listening_histories}
+  result = []
+  for e in episodes:
+    try:
+      current_progress = episode_update_info_map[e.id]['current_progress']
+      percentage_listened = episode_update_info_map[e.id]['percentage_listened']
+    except KeyError:
+      raise Exception('Invalid episode_id {} provided'.format(e.id))
```

Comment:
Should an exception be raised, here? Or maybe it should try to create/update all the episodes that have a valid episode_id?

Reply:
Yeah idk what best practices for batches are, but I feel like if some part of the batch is invalid it might be better to not create anything because episode_id would only be invalid if there was a bug on client